File: fe2-android/app/src/test/unit/java/com/github/dedis/popstellar/repository/database/SocialMediaDatabaseTest.java
package com.github.dedis.popstellar.repository.database;

import androidx.test.core.app.ApplicationProvider;
import androidx.test.ext.junit.runners.AndroidJUnit4;

import com.github.dedis.popstellar.di.AppDatabaseModuleHelper;
import com.github.dedis.popstellar.model.objects.*;
import com.github.dedis.popstellar.model.objects.security.MessageID;
import com.github.dedis.popstellar.model.objects.security.PublicKey;
import com.github.dedis.popstellar.repository.database.socialmedia.*;

import org.junit.*;
import org.junit.runner.RunWith;

import java.time.Instant;
import java.util.List;

import io.reactivex.observers.TestObserver;

import static com.github.dedis.popstellar.testutils.Base64DataUtils.generateMessageID;
import static com.github.dedis.popstellar.testutils.Base64DataUtils.generatePublicKey;

@RunWith(AndroidJUnit4.class)
public class SocialMediaDatabaseTest {

  private static AppDatabase appDatabase;
  private static ChirpDao chirpDao;
  private static ReactionDao reactionDao;

  private static final String LAO_ID = Lao.generateLaoId(generatePublicKey(), 1000, "LAO");

  private static final PublicKey SENDER = generatePublicKey();
  private static final MessageID CHIRP1_ID = generateMessageID();
  private static final MessageID CHIRP2_ID = generateMessageID();
  private static final String EMOJI = "\uD83D\uDC4D";
  private static final Chirp CHIRP_1 =
      new Chirp(CHIRP1_ID, SENDER, "This is a chirp !", 1001, new MessageID(""));
  private static final Chirp CHIRP_2 =
      new Chirp(CHIRP2_ID, SENDER, "This is another chirp !", 1003, new MessageID(""));

  private static final Reaction REACTION_1 =
      new Reaction(generateMessageID(), SENDER, EMOJI, CHIRP1_ID, Instant.now().getEpochSecond());

  private static final ChirpEntity CHIRP_ENTITY = new ChirpEntity(LAO_ID, CHIRP_1);
  private static final ReactionEntity REACTION_ENTITY = new ReactionEntity(REACTION_1);

  @Before
  public void before() {
    appDatabase =
        AppDatabaseModuleHelper.getAppDatabase(ApplicationProvider.getApplicationContext());
    chirpDao = appDatabase.chirpDao();
    reactionDao = appDatabase.reactionDao();
  }

  @After
  public void close() {
    appDatabase.close();
  }

  @Test
  public void insertChirpTest() {
    TestObserver<Void> testObserver = chirpDao.insert(CHIRP_ENTITY).test();

    testObserver.awaitTerminalEvent();
    testObserver.assertComplete();
  }

  @Test
  public void insertAndGetChirpTest() {
    TestObserver<Void> testObserver = chirpDao.insert(CHIRP_ENTITY).test();

    testObserver.awaitTerminalEvent();
    testObserver.assertComplete();

    TestObserver<List<Chirp>> testObserver2 =
        chirpDao
            .getChirpsByLaoId(LAO_ID)
            .test()
            .assertValue(chirps -> chirps.size() == 1 && chirps.get(0).equals(CHIRP_1));

    testObserver2.awaitTerminalEvent();
    testObserver2.assertComplete();
  }

  @Test
  public void insertReactionTest() {
    TestObserver<Void> testObserver = reactionDao.insert(REACTION_ENTITY).test();

    testObserver.awaitTerminalEvent();
    testObserver.assertComplete();
  }

  @Test
  public void insertAndGetReactionTest() {
    TestObserver<Void> testObserver = reactionDao.insert(REACTION_ENTITY).test();

    testObserver.awaitTerminalEvent();
    testObserver.assertComplete();

    TestObserver<List<Reaction>> testObserver2 =
        reactionDao
            .getReactionsByChirpId(CHIRP1_ID)
            .test()
            .assertValue(reactions -> reactions.size() == 1 && reactions.get(0).equals(REACTION_1));

    testObserver2.awaitTerminalEvent();
    testObserver2.assertComplete();
  }
}

File: fe2-android/app/src/test/unit/java/com/github/dedis/popstellar/repository/database/SocialMediaDatabaseTest.kt
package com.github.dedis.popstellar.repository.database

import androidx.test.core.app.ApplicationProvider
import androidx.test.ext.junit.runners.AndroidJUnit4
import com.github.dedis.popstellar.di.AppDatabaseModuleHelper.getAppDatabase
import com.github.dedis.popstellar.model.objects.Chirp
import com.github.dedis.popstellar.model.objects.Lao.Companion.generateLaoId
import com.github.dedis.popstellar.model.objects.Reaction
import com.github.dedis.popstellar.model.objects.security.MessageID
import com.github.dedis.popstellar.repository.database.socialmedia.ChirpDao
import com.github.dedis.popstellar.repository.database.socialmedia.ChirpEntity
import com.github.dedis.popstellar.repository.database.socialmedia.ReactionDao
import com.github.dedis.popstellar.repository.database.socialmedia.ReactionEntity
import com.github.dedis.popstellar.testutils.Base64DataUtils
import io.reactivex.observers.TestObserver
import java.time.Instant
import org.junit.After
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4::class)
class SocialMediaDatabaseTest {
  private lateinit var appDatabase: AppDatabase
  private lateinit var chirpDao: ChirpDao
  private lateinit var reactionDao: ReactionDao

  @Before
  fun before() {
    appDatabase = getAppDatabase(ApplicationProvider.getApplicationContext())
    chirpDao = appDatabase.chirpDao()
    reactionDao = appDatabase.reactionDao()
  }

  @After
  fun close() {
    appDatabase.close()
  }

  @Test
  fun insertChirpTest() {
    val testObserver = chirpDao.insert(CHIRP_ENTITY).test()
    testObserver.awaitTerminalEvent()
    testObserver.assertComplete()
  }

  @Test
  fun insertAndGetChirpTest() {
    val testObserver = chirpDao.insert(CHIRP_ENTITY).test()
    testObserver.awaitTerminalEvent()
    testObserver.assertComplete()

    val testObserver2: TestObserver<List<Chirp>?> =
      chirpDao.getChirpsByLaoId(LAO_ID).test().assertValue { chirps: List<Chirp> ->
        chirps.size == 1 && chirps[0] == CHIRP_1
      }
    testObserver2.awaitTerminalEvent()
    testObserver2.assertComplete()
  }

  @Test
  fun insertReactionTest() {
    val testObserver = reactionDao.insert(REACTION_ENTITY).test()
    testObserver.awaitTerminalEvent()
    testObserver.assertComplete()
  }

  @Test
  fun insertAndGetReactionTest() {
    val testObserver = reactionDao.insert(REACTION_ENTITY).test()
    testObserver.awaitTerminalEvent()
    testObserver.assertComplete()
    val testObserver2: TestObserver<List<Reaction>?> =
      reactionDao.getReactionsByChirpId(CHIRP1_ID).test().assertValue { reactions: List<Reaction> ->
        reactions.size == 1 && reactions[0] == REACTION_1
      }
    testObserver2.awaitTerminalEvent()
    testObserver2.assertComplete()
  }

  companion object {
    private val LAO_ID = generateLaoId(Base64DataUtils.generatePublicKey(), 1000, "LAO")
    private val SENDER = Base64DataUtils.generatePublicKey()
    private val CHIRP1_ID = Base64DataUtils.generateMessageID()
    private val CHIRP2_ID = Base64DataUtils.generateMessageID()
    private const val EMOJI = "\uD83D\uDC4D"
    private val CHIRP_1 = Chirp(CHIRP1_ID, SENDER, "This is a chirp !", 1001, MessageID(""))
    private val CHIRP_2 = Chirp(CHIRP2_ID, SENDER, "This is another chirp !", 1003, MessageID(""))
    private val REACTION_1 =
      Reaction(
        Base64DataUtils.generateMessageID(),
        SENDER,
        EMOJI,
        CHIRP1_ID,
        Instant.now().epochSecond
      )
    private val CHIRP_ENTITY = ChirpEntity(LAO_ID, CHIRP_1)
    private val REACTION_ENTITY = ReactionEntity(REACTION_1)
  }
}
